### PR TITLE
Bug 1216523 - Fix task manager enter/exit transitions. r=etienne

### PR DESCRIPTION
**Pull request by Cwiiis**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=80d87a215bb0228557fa475aa29562934778d063

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=6d2bcb3541f9429a26ec6ecb552c41417dabcafd

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b7d068c477941f2c2c8b1ce5c2a0a2e32e25bc90

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c5415a1ae134a01199ff8ecabd80f6d028127818

File: apps/system/js/app_window.js [hunk 1]
```diff
@@ -2126,14 +2126,22 @@
...
+      this.debug('requesting, then showing screenshot for cardsview.');
+      this.getScreenshot(this._showScreenshotOverlay.bind(this));
+    } else {
+      this.debug('showing screenshot for cardsview.');
+      this._showScreenshotOverlay();
+    }
   };
 
   AppWindow.prototype._handle__cardviewshown = function aw_cvshown() {
+    this._showScreenshotOverlay();
```
Comment (etiennesegonzac) [inline apps/system/js/app_window.js]: needed? we always trigger it in `beforeshow`

File: apps/system/js/app_window.js [hunk 2]
```diff
@@ -2126,14 +2126,22 @@
...
     if (this.element && this.element.classList.contains('no-screenshot') &&
         this._screenshotBlob) {
       this.element.classList.remove('no-screenshot');
+      this.screenshotOverlay.classList.add('visible');
```
Comment (etiennesegonzac) [inline apps/system/js/app_window.js]: needed? `_showScreenshotOverlay` will do it itself

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7c1ef0279629d62f34ce9c9f2551ca858bf0cd71

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=38f2b226d78bd180cbfdc9358b12daa3dce6c1fe

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=db82ead09d0c5bb2f5e7e22030433e73e16e20e8

File: apps/system/js/app_window.js (hunk 1 above)
Comment (Cwiiis) [inline apps/system/js/app_window.js]: This guarantees the trigger (in case this.getScreenshot takes a long time to complete).

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=1bc9bc37df151db21b0621e97bb393f64e1517d4

File: apps/system/js/task_manager_utils.js [hunk 3]
```diff
@@ -45,34 +45,13 @@ exports.TaskManagerUtils = {
         icon.renderBlob(iconObject.blob, { size: size });
       }
     }, (err) => {
+      element.classList.remove('pending');
       console.warn('getSiteIconUrl failed to resolve an icon:', err.message);
     }).then(() => {
       element.classList.remove('pending');
```
Comment (etiennesegonzac) [inline apps/system/js/task_manager_utils.js]: I'm still getting a constant flash of the "gray rocket icon" before the icon is displayed, this seems to help
```diff
diff --git a/apps/system/js/task_manager_utils.js b/apps/system/js/task_manager_utils.js
index 3ceb7ce..6803aec 100644
--- a/apps/system/js/task_manager_utils.js
+++ b/apps/system/js/task_manager_utils.js
@@ -37,17 +37,19 @@ exports.TaskManagerUtils = {
 
     return url;
   },
 
   loadAppIcon(app, element, size) {
     return app.getSiteIconUrl(size).then((iconObject) => {
       if (iconObject && iconObject.originalUrl) {
         var icon = new Icon(element, iconObject.originalUrl);
-        icon.renderBlob(iconObject.blob, { size: size });
+        return new Promise(function(resolve) {
+          icon.renderBlob(iconObject.blob, { size: size, onLoad: resolve });
+        });
       }
     }, (err) => {
       element.classList.remove('pending');
       console.warn('getSiteIconUrl failed to resolve an icon:', err.message);
     }).then(() => {
       element.classList.remove('pending');
     });
   },

```

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c11b3c6abdc3f05540796296f655f068e3672b86

File: apps/system/js/task_manager_utils.js (hunk 3 above)
Comment (Cwiiis) [inline apps/system/js/task_manager_utils.js]: Good call :)

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=12be8607daf871ea557cc1a77b6ee570158d6dd3

**Pull request merged**